// @tsed/di/lib/esm/common/utils/providerBuilder.js
import "../registries/ProviderRegistry.js";
import { ProviderType } from "../domain/ProviderType.js";
import { injector } from "../fn/injector.js";
import { GlobalProviders } from "../registries/GlobalProviders.js";
export function providerBuilder(props, baseOpts = {}) {
    return (token, options = {}) => {
        const merged = {
            global: !injector().isLoaded(),
            ...options,
            ...baseOpts,
            token
        };
        const provider = GlobalProviders.merge(token, merged);
        if (!merged.global) {
            injector().setProvider(token, provider);
        }
        return props.reduce((acc, prop) => {
            return {
                ...acc,
                [prop]: function (value) {
                    provider[prop] = value;
                    return this;
                }
            };
        }, {
            factory(factory) {
                provider.reset();
                provider.useFactory = factory;
                return this;
            },
            asyncFactory(asyncFactory) {
                provider.reset();
                provider.useAsyncFactory = asyncFactory;
                return this;
            },
            value(value) {
                provider.reset();
                provider.useValue = value;
                provider.type = ProviderType.VALUE;
                return this;
            },
            class(k) {
                provider.reset();
                provider.useClass = k;
                return this;
            },
            store() {
                return provider.store;
            },
            inspect() {
                return provider;
            },
            token() {
                return provider.token;
            },
            set(key, value) {
                this.store().set(key, value);
                return this;
            }
        });
    };
}
