// @tsed/platform-http/lib/esm/common/utils/getBestHost.js
import { getHostInfoFromPort } from "@tsed/core";
import { configuration, constant } from "@tsed/di";
function httpPort() {
    return constant("httpPort", constant("port"));
}
export function setHttpPort(settings) {
    configuration().set("httpPort", `${settings.address}:${settings.port}`);
}
export function setHttpsPort(settings) {
    configuration().set("httpsPort", `${settings.address}:${settings.port}`);
}
export function getHttpsPort() {
    return getHostInfoFromPort("https", constant("httpsPort"));
}
export function getHttpPort() {
    return getHostInfoFromPort("http", httpPort());
}
export function getBestHost() {
    if (constant("httpsPort")) {
        return getHttpsPort();
    }
    if (httpPort()) {
        return getHttpPort();
    }
    return {
        toString() {
            return "/";
        }
    };
}
configuration().decorate("getBestHost", getBestHost);
configuration().decorate("getHttpPort", getHttpPort);
configuration().decorate("getHttpsPort", getHttpsPort);
configuration().decorate("setHttpsPort", setHttpsPort);
configuration().decorate("setHttpPort", setHttpPort);
