// @tsed/platform-http/lib/esm/testing/FakeAdapter.js
import { injectable } from "@tsed/di";
import { PlatformAdapter } from "../common/index.js";
export class FakeAdapter extends PlatformAdapter {
    constructor() {
        super(...arguments);
        this.NAME = "FAKE_ADAPTER";
        this.providers = [];
    }
    static createFakeRawDriver() {
        // istanbul ignore next
        function FakeRawDriver() { }
        // istanbul ignore next
        function use() {
            return this;
        }
        FakeRawDriver.use = use;
        FakeRawDriver.all = use;
        FakeRawDriver.get = use;
        FakeRawDriver.patch = use;
        FakeRawDriver.post = use;
        FakeRawDriver.put = use;
        FakeRawDriver.head = use;
        FakeRawDriver.delete = use;
        FakeRawDriver.options = use;
        return FakeRawDriver;
    }
    createApp() {
        const app = FakeAdapter.createFakeRawDriver();
        return {
            app,
            callback() {
                return app;
            }
        };
    }
    multipart(options) {
        return {};
    }
    statics(endpoint, options) {
        return {};
    }
    bodyParser(type) {
        return () => { };
    }
    mapLayers(layers) { }
    mapHandler(handler) {
        return handler;
    }
    useContext() { }
}
injectable(PlatformAdapter).class(FakeAdapter);
