// @tsed/di/lib/esm/common/domain/Provider.js
import { Store } from "@tsed/core/types/Store.js";
import { classOf } from "@tsed/core/utils/classOf.js";
import { getClassOrSymbol } from "@tsed/core/utils/getClassOrSymbol.js";
import { isClass } from "@tsed/core/utils/isClass.js";
import { nameOf } from "@tsed/core/utils/nameOf.js";
import { DI_USE_PARAM_OPTIONS } from "../constants/constants.js";
import { discoverHooks } from "../utils/discoverHooks.js";
import { ProviderScope } from "./ProviderScope.js";
import { ProviderType } from "./ProviderType.js";
export class Provider {
    #useClass;
    #token;
    #store;
    #tokenStore;
    constructor(token, options = {}) {
        /**
         * Token group provider to retrieve all provider from the same type
         */
        this.type = ProviderType.PROVIDER;
        this.hooks = {};
        this.token = token;
        this.useClass = token;
        Object.assign(this, options);
        if (options instanceof Provider) {
            this.#useClass = options.#useClass;
            this.#store = options.#store;
            this.#tokenStore = options.#tokenStore;
        }
    }
    get token() {
        return this.#token;
    }
    set token(value) {
        if (value) {
            this.#token = getClassOrSymbol(value);
            this.#tokenStore = this.#store = Store.from(value);
        }
    }
    /**
     * @deprecated
     */
    get provide() {
        return this.token;
    }
    /**
     * @deprecated
     * @param value
     */
    set provide(value) {
        this.token = value;
    }
    get useClass() {
        return this.#useClass;
    }
    /**
     * Create a new store if the given value is a class. Otherwise, the value is ignored.
     * @param value
     */
    set useClass(value) {
        if (isClass(value)) {
            this.#useClass = classOf(value);
            this.#store = Store.from(value);
            this.hooks = discoverHooks(this.#useClass);
        }
    }
    get className() {
        return this.name;
    }
    get name() {
        return nameOf(this.token);
    }
    get store() {
        return this.#store;
    }
    get path() {
        return this.store.get("path", "/");
    }
    set path(path) {
        this.store.set("path", path);
    }
    /**
     * Get the scope of the provider.
     *
     * ::: tip Note
     * Async provider is always a SINGLETON
     * :::
     *
     * @returns {boolean}
     */
    get scope() {
        if (this.isAsync()) {
            return ProviderScope.SINGLETON;
        }
        return this.get("scope", ProviderScope.SINGLETON);
    }
    /**
     * Change the scope value of the provider.
     * @param scope
     */
    set scope(scope) {
        this.store.set("scope", scope);
    }
    get configuration() {
        return this.get("configuration");
    }
    set configuration(configuration) {
        this.store.set("configuration", configuration);
    }
    get children() {
        return this.store.get("childrenControllers", []);
    }
    set children(children) {
        this.store.set("childrenControllers", children);
    }
    getArgOpts(index) {
        return this.store.get(`${DI_USE_PARAM_OPTIONS}:${index}`);
    }
    get(key, defaultValue) {
        return this.store.get(key) || this.#tokenStore.get(key) || defaultValue;
    }
    isAsync() {
        return !!this.useAsyncFactory;
    }
    clone() {
        return new (classOf(this))(this.token, this);
    }
    /**
     *
     * @returns {boolean}
     */
    hasChildren() {
        return !!this.children.length;
    }
    /**
     *
     * @returns {boolean}
     */
    hasParent() {
        return !!this.store.get("parentController");
    }
    toString() {
        return [
            "Token",
            this.name,
            this.useClass && nameOf(this.useClass),
            this.useFactory && "Factory",
            this.useValue && "Value",
            this.useAsyncFactory && "AsyncFactory"
        ]
            .filter(Boolean)
            .join(":");
    }
    reset() {
        this.useValue = undefined;
        this.useFactory = undefined;
        this.useAsyncFactory = undefined;
        return this;
    }
}
